src/c-writer.cc

```diff
@@ -1026,12 +1026,6 @@ void CWriter::WriteTables() {
     return;
   }
 
-  if (!module_->types.size()) {
-    // If no types are defined then there is no way to use the table
-    // for anything.
-    return;
-  }
```

Comment:
Hmm.. why was this needed?

Reply:
Here's a test case that fails without this change:
```wat
(module (table 0 funcref) (export "a" (table 0)))
```

The module doesn't have any types, so, this code causes `CWriter::WriteTables` to early-out and omit the call to `WriteTable(DefineGlobalScopeName(table->name));`.

Later, when `CWriter::WriteExports(WriteExportsKind::Initializers)` is called, it's trying to write the body of the `init_exports` function, and it tries to execute `Write(mangled_name, " = ", ExternalPtr(internal_name), ";");` (https://github.com/WebAssembly/wabt/blob/main/src/c-writer.cc#L1234), and the call to `Write(const ExternalPtr&)` tries to call `GetGlobalName` which hits the assertion failure (this one: https://github.com/WebAssembly/wabt/blob/main/src/c-writer.cc#L595) because the table's name never made it to `DefineGlobalScopeName` back when `WriteTables` was called.

I guess the big-picture subtlety here is that these `CWriter` methods are designed to be invoked in a particular order, in which some of them call `DefineGlobalScopeName` (the first time a name is seen) and then other methods call `GetGlobalName` on the same name and assume that the name has already been defined. And these ordering requirements/dependencies between the methods are not completely obvious or explicit.

Reply:
Oh, yes good point.  That looks like a fix for a real issue.